### PR TITLE
WRP-3673: Add a pack option `--content-hash` to support better caching

### PR DESCRIPTION
**Pull request by hong6316**
### Checklist

* [x] I have read and understand the [contribution guide](http://enactjs.com/docs/developer-guide/contributing/)
* [x] A [CHANGELOG entry](http://enactjs.com/docs/developer-guide/contributing/changelogs/) is included
* [ ] At least one test case is included for this feature or bug fix
* [x] Documentation was added or is not needed
* [ ] This is an API breaking change

### Issue Resolved / Feature Added
[//]: # (Describe the issue resolved or feature added by this pull request)
Some Enact apps are running as hosted web app. We need to support a way for the client can update hosted app files if the app is updated so that the client doesn't stay with the local cache.

One of the possible candidates is to use webpack hash config when pack.
(Caching description: https://webpack.kr/guides/caching)
(Set output description: https://webpack.js.org/configuration/output/#outputfilename, https://webpack.js.org/configuration/output/#outputchunkfilename)

### Resolution
[//]: # (Does the code work as intended?)
[//]: # (What is the impact of this change and *why* was it made?)
- Add a pack option `--content-hash` to be used by the hosted web app
- If the option is true, add a unique hash value based on content to the output file name so that the changed files could be loaded again.

### Additional Considerations
[//]: # (How should the change be tested?)
[//]: # (Are there any outstanding questions?)
[//]: # (Were any side-effects caused by the change?)

### Links
[//]: # (Related issues, references)
WRP-3673

### Comments
Enact-DCO-1.0-Signed-off-by: Taeyoung Hong (taeyoung.hong@lge.com)

### CODE REVIEW

**Comment by MikyungKim** [comment]
Please provide which documentation you referred to make this PR.

File: commands/pack.js [hunk 1]
```diff
@@ -62,6 +62,7 @@ function displayHelp() {
 			--externals-polyfill  	Flag whether to use external polyfill (or include in framework build)
 			--ilib-additional-path	Specify iLib additional resources path
 			--no-animation          Build without effects such as animation and shadow
+			--add-contenthash       Add the hash of the chunk, including only elements of this content type
```
Comment (MikyungKim) [inline commands/pack.js:65]: For what reason, does this need to be a private option?
If not, please make it public.

Comment (MikyungKim) [reply commands/pack.js:65]: Let's use `--content-hash` for the option name.

File: config/webpack.config.js [hunk 2]
```diff
@@ -210,7 +216,7 @@ module.exports = function (env, isomorphic = false, noAnimation = false, ilibAdd
 			// We don't currently advertise code splitting but Webpack supports it.
 			filename: '[name].js',
 			// There are also additional JS chunk files if you use code splitting.
-			chunkFilename: 'chunk.[name].js',
+			chunkFilename: contenthash ? 'chunk.[name].[contenthash].js' : 'chunk.[name].js',
```
Comment (MikyungKim) [inline config/webpack.config.js:219]: Why did you add this only to chunk files?
I think we could use it to `filename` also.
How about assets like image files?

File: commands/pack.js (hunk 1 above)
Comment (seunghoh) [reply commands/pack.js:65]: Let's make the option public and add docs to explain.

Comment (seunghoh) [reply commands/pack.js:65]: `Add the hash of the chunk, including only elements of this content type` sounds not straightforward.

File: config/webpack.config.js (hunk 2 above)
Comment (hong6316) [reply config/webpack.config.js:219]: For several reasons, the filename also needs to be changed.(image file, apply main's changes, etc.)
I will change this.

File: commands/pack.js [hunk 3]
```diff
@@ -47,6 +47,7 @@ function displayHelp() {
 	console.log('                      (requires V8_MKSNAPSHOT set)');
 	console.log('    -m, --meta        JSON to override package.json enact metadata');
 	console.log('    -c, --custom-skin Build with a custom skin');
+	console.log('    --content-hash    Add the hash of the chunk');
```
Comment (MikyungKim) [inline commands/pack.js:50]:
```suggestion
	console.log('    --content-hash    Add a unique hash to output files based on the content of an asset');
```

File: config/webpack.config.js [hunk 4]
```diff
@@ -41,7 +41,13 @@ const createEnvironmentHash = require('./createEnvironmentHash');
 
 // This is the production and development configuration.
 // It is focused on developer experience, fast rebuilds, and a minimal bundle.
-module.exports = function (env, isomorphic = false, noAnimation = false, ilibAdditionalResourcesPath) {
+module.exports = function (
+	env,
+	isomorphic = false,
+	noAnimation = false,
+	contenthash = false,
```
Comment (MikyungKim) [inline config/webpack.config.js:48]:
```suggestion
	contentHash = false,
```

File: docs/building-apps.md [hunk 5]
```diff
@@ -23,6 +23,7 @@ order: 4
     -s, --snapshot    Generate V8 snapshot blob
                       (requires V8_MKSNAPSHOT set)
     -c, --custom-skin Build with a custom skin
+    --content-hash    Add the hash of the chunk
```
Comment (MikyungKim) [inline docs/building-apps.md:26]:
```suggestion
    --content-hash    Add a unique hash to output files based on the content of an asset
```

File: docs/building-apps.md [hunk 6]
```diff
@@ -218,6 +219,25 @@ my-app/
   webos-meta/
 ```
 
+## [contenthash] Injection
+
+For supporting [`Caching`](https://webpack.kr/guides/caching), Enact CLI supports adding the hash of the chunk. You can use the [`contenthash`](https://webpack.js.org/configuration/output/#template-strings) substitutions setting to define the names of your output files.
```
Comment (MikyungKim) [inline docs/building-apps.md:224]: Don't use kr domain address, please.

File: commands/pack.js (hunk 3 above)
Comment (MikyungKim) [reply commands/pack.js:50]: Let's add `-h` for shorthand.

Comment (MikyungKim) [reply commands/pack.js:50]: And move this option order to after `output`.

File: commands/pack.js [hunk 7]
```diff
@@ -255,6 +256,7 @@ function api(opts = {}) {
 		opts.production ? 'production' : 'development',
 		opts.isomorphic,
 		!opts.animation,
+		opts['content-hash'],
```
Comment (MikyungKim) [inline commands/pack.js:259]:
```suggestion
		opts['content-hash'],
		opts.isomorphic,
		!opts.animation,
```

File: commands/pack.js [hunk 8]
```diff
@@ -291,6 +293,7 @@ function cli(args) {
 			'isomorphic',
 			'snapshot',
 			'animation',
+			'content-hash',
```
Comment (MikyungKim) [inline commands/pack.js:296]: Let's move it to the top of the array.

File: config/webpack.config.js (hunk 2 above)
Comment (MikyungKim) [reply config/webpack.config.js:219]: Where is the change for "assets" like image files?

File: docs/building-apps.md [hunk 9]
```diff
@@ -218,6 +219,25 @@ my-app/
   webos-meta/
 ```
 
+## [contenthash] Injection
```
Comment (MikyungKim) [inline docs/building-apps.md:222]: This documentation describes not the `content-hash` option, but the inside code. Don't you think?

Comment (MikyungKim) [inline docs/building-apps.md:222]:
```suggestion
## Caching Support
```

File: docs/building-apps.md (hunk 6 above)
Comment (MikyungKim) [inline docs/building-apps.md:224]: I can't see `content-hash` option in this document. Please re-write it.

**Review by MikyungKim — COMMENTED** [review]
Please add a changelog for this.

File: config/webpack.config.js (hunk 2 above)
Comment (hong6316) [reply config/webpack.config.js:219]: I just add it too.
(verification result: check WRP-3673)

File: docs/building-apps.md (hunk 6 above)
Comment (hong6316) [reply docs/building-apps.md]: OK

File: docs/building-apps.md (hunk 9 above)
Comment (hong6316) [reply docs/building-apps.md]: OK

File: commands/pack.js (hunk 8 above)
Comment (hong6316) [reply commands/pack.js]: OK

File: commands/pack.js (hunk 7 above)
Comment (hong6316) [reply commands/pack.js]: OK

File: docs/building-apps.md (hunk 6 above)
Comment (hong6316) [reply docs/building-apps.md]: OK

File: docs/building-apps.md (hunk 5 above)
Comment (hong6316) [reply docs/building-apps.md]: OK

File: config/webpack.config.js (hunk 4 above)
Comment (hong6316) [reply config/webpack.config.js]: OK

File: commands/pack.js (hunk 3 above)
Comment (hong6316) [reply commands/pack.js]: OK

File: CHANGELOG.md [hunk 10]
```diff
@@ -1,5 +1,8 @@
 ## unreleased
 
+### pack
```
Comment (MikyungKim) [inline CHANGELOG.md:3]: It should go to the above of `### serve`.

File: docs/building-apps.md [hunk 11]
```diff
@@ -8,22 +8,23 @@ order: 4
...
             "signage" - Locales supported on webOS signage
             "all" - All locales that iLib supports
-    -s, --snapshot    Generate V8 snapshot blob
-                      (requires V8_MKSNAPSHOT set)
-    -c, --custom-skin Build with a custom skin
-    --stats           Output bundle analysis file
+    -s, --snapshot       Generate V8 snapshot blob
+                         (requires V8_MKSNAPSHOT set)
+    -c, --custom-skin    Build with a custom skin
+    -ch, --content-hash  Add the hash of the chunk
```
Comment (MikyungKim) [inline docs/building-apps.md:26]: Please update the description with the latest one.
Let's move it to the top.

File: commands/pack.js [hunk 12]
```diff
@@ -304,6 +307,7 @@ function cli(args) {
 			l: 'locales',
 			s: 'snapshot',
 			m: 'meta',
+			ch: 'content-hash',
```
Comment (hong6316) [inline commands/pack.js:310]: This alias does not work.

File: docs/building-apps.md (hunk 11 above)
Comment (hong6316) [reply docs/building-apps.md]: OK

File: commands/pack.js [hunk 13]
```diff
@@ -31,6 +31,7 @@ function displayHelp() {
 	console.log();
 	console.log('  Options');
 	console.log('    -o, --output      Specify an output directory');
+	console.log('    --content-hash    Add a unique hash to output files based on the content of an asset');
```
Comment (MikyungKim) [inline commands/pack.js:34]: Let's make this clearer.
```suggestion
	console.log('    --content-hash    Add a unique hash to output file names based on the content of an asset');
```

File: docs/building-apps.md [hunk 14]
```diff
@@ -8,6 +8,7 @@ order: 4
     enact pack [options]
 
   Options
+    --content-hash    Add a unique hash to output files based on the content of an asset
```
Comment (MikyungKim) [inline docs/building-apps.md:11]: here also.
```suggestion
    --content-hash    Add a unique hash to output file names based on the content of an asset
```

File: docs/building-apps.md [hunk 15]
```diff
@@ -218,6 +219,25 @@ my-app/
...
 
+## Caching
+
+For supporting better [`Caching`](https://webpack.js.org/guides/caching/), Enact CLI provides `--content-hash` option to add a unique hash to each output file name based on the content of an asset.
+
+Building With this option should produce the following output:
+
+```none
+                       Asset       Size  Chunks                    Chunk Names
+main.7e2c49a622975ebd9b7e.js     544 kB       0  [emitted]  [big]  main
```
Comment (MikyungKim) [inline docs/building-apps.md:230]: Please provide an example with css files and image files too.

Comment (hong6316) [reply docs/building-apps.md]: OK, 
image files are not shown on bash, so I will update css only

File: docs/building-apps.md [hunk 16]
```diff
@@ -218,6 +219,25 @@ my-app/
...
+        199.85 kB       dist/main.2088c66150ab73b27793.css
+```
+
+When the content changes, the output file name will change as well.
+```none
+        1.11 MB         dist/main.7544f55b64439c8d0f0e.js
+        199.85 kB       dist/main.2088c66150ab73b27793.css
+```
+
+> **NOTE** The filename `main.*.js` will be changed after bundling, without changes. This is because webpack includes certain boilerplate, specifically the runtime and manifest, in the entry chunk.
```
Comment (MikyungKim) [inline docs/building-apps.md:239]:
```suggestion
> **NOTE** The filename `main.*.js` will be changed after another building, even without making any changes. This is because webpack includes certain boilerplate, specifically the runtime and manifest, in the entry chunk.
```

File: docs/building-apps.md [hunk 17]
```diff
@@ -218,6 +219,25 @@ my-app/
   webos-meta/
 ```
 
+## Caching
+
+For supporting better [`Caching`](https://webpack.js.org/guides/caching/), Enact CLI provides `--content-hash` option to add a unique hash to each output file name based on the content of an asset.
```
Comment (MikyungKim) [inline docs/building-apps.md:224]:
```suggestion
For supporting better [`caching`](https://webpack.js.org/guides/caching/), Enact CLI provides `--content-hash` option to add a unique hash to each output file name based on the content of an asset.
```

**Review by MikyungKim — APPROVED** [review]
LGTM
